dhall/src/Dhall/Binary.hs

```diff
@@ -105,860 +116,1108 @@ unApply e₀ = (baseFunction₀, diffArguments₀ [])
 
-            let x₁  = TString x₀
+            TypeUInt64 -> do
+                !n <- Decoding.decodeWord64
 
-            let _T₁ = case mT₀ of
-                    Nothing  -> TNull
-                    Just _T₀ -> encode _T₀
+                return (Var (V "_" (fromIntegral n)))
 
-            return (x₁, _T₁)
-    encode (BoolLit b) =
-        TBool b
-    encode (BoolIf t₀ l₀ r₀) =
-        TList [ TInt 14, t₁, l₁, r₁ ]
-      where
-        t₁ = encode t₀
-        l₁ = encode l₀
-        r₁ = encode r₀
-    encode (NaturalLit n) =
-        TList [ TInt 15, TInteger (fromIntegral n) ]
-    encode (IntegerLit n) =
-        TList [ TInt 16, TInteger n ]
-    encode (DoubleLit d) =
-        encode d
-    encode (TextLit (Chunks xys₀ z₀)) =
-        TList ([ TInt 18 ] ++ xys₁ ++ [ z₁ ])
-      where
-        xys₁ = do
-            (x₀, y₀) <- xys₀
-            let x₁ = TString x₀
-            let y₁ = encode y₀
-            [ x₁, y₁ ]
-
-        z₁ = TString z₀
-    encode (Assert t₀) =
-        TList [ TInt 19, t₁ ]
-      where
-        t₁ = encode t₀
-    encode (Embed x) =
-        encode x
-    encode (Let a b) =
-        TList ([ TInt 25 ] ++ as₁ ++ [ b₁ ])
-      where
-        MultiLet as₀ b₀ = Dhall.Syntax.multiLet a b
+            TypeFloat16 -> do
+                !n <- Decoding.decodeFloat
 
-        as₁ = do
-            Binding _ x₀ _ mA₀ _ a₀ <- toList as₀
+                return (DoubleLit (DhallDouble (float2Double n)))
 
-            let mA₁ = case mA₀ of
-                    Nothing       -> TNull
-                    Just (_, _A₀) -> encode _A₀
+            TypeFloat32 -> do
+                !n <- Decoding.decodeFloat
 
-            let a₁ = encode a₀
+                return (DoubleLit (DhallDouble (float2Double n)))
 
-            [ TString x₀, mA₁, a₁ ]
+            TypeFloat64 -> do
+                !n <- Decoding.decodeDouble
 
-        b₁ = encode b₀
-    encode (Annot t₀ _T₀) =
-        TList [ TInt 26, t₁, _T₁ ]
-      where
-        t₁  = encode t₀
-        _T₁ = encode _T₀
-    encode (ToMap t₀ Nothing) =
-        TList [ TInt 27, t₁ ]
-      where
-        t₁ = encode t₀
-    encode (ToMap t₀ (Just _T₀)) =
-        TList [ TInt 27, t₁, _T₁ ]
-      where
-        t₁  = encode t₀
-        _T₁ = encode _T₀
-    encode (Note a _) = absurd a
-
-instance ToTerm Import where
-    encode import_ =
-        case importType of
-            Remote (URL { scheme = scheme₀, ..}) ->
-                TList
-                    (   prefix
-                    ++  [ TInt scheme₁, using, TString authority ]
-                    ++  map TString (reverse components)
-                    ++  [ TString file ]
-                    ++  (case query    of Nothing -> [ TNull ]; Just q -> [ TString q ])
+                return (DoubleLit (DhallDouble n))
+
+            TypeBool -> do
+                !b <- Decoding.decodeBool
+
+                return (BoolLit b)
+
+            TypeString -> do
+                s <- Decoding.decodeString
```

Comment:
Can you add a type annotation that shows that this is a `Text`? `decodeString` is pretty confusing! :/

Reply:
There are a lot of `decodeString`s that would need to be changed if we did that

Reply:
Right. Never mind then!